docs/source/cuda/cuda_array_interface.rst

```diff
@@ -77,7 +77,101 @@ The following are optional entries:
+synchronization of the interface:
+
+- If a device array is bound to a stream, then when the interface is exported
+  (i.e. at the time the ``__cuda_array_interface__`` property of a device array
+  is accessed), Numba will record an event in the array's stream and insert a
+  wait on the event in the default stream.
+- If a device array is imported (e.g. using :func:`cuda.as_cuda_array` on an
+  object providing ``__cuda_array_interface__``, then Numba will bind the
+  imported array to the default stream so that any operations on it will be
+  synchronized correctly even if the user does nothing to bind the array to a
+  stream.
+- If the synchronization on the default stream when importing is undesired (e.g.
+  in the scenario described above where no synchronization is necessary) then
+  the binding of the array to the default stream can be avoided by passing
+  ``sync=False`` to either of :func:`cuda.from_cuda_array_interface` or
+  :func:`as_cuda_array`.
+
+This means that:
+
+- If the user is only operating on the default stream, then no further action on
+  the part of the user is required.
+- If the user performs operations on non-default streams, then device arrays
+  they use that may be exported to other frameworks should be operated on in the
+  stream that they are bound to - this binding can take place either at
+  construction time, or by using the
+  :func:`numba.cuda.cudadrv.devicearray.DeviceNDArray.bind` method.
+- If synchronization is to be avoided on both import and export, then device
+  arrays should not be bound to a particular stream (instead the stream must be
+  specified for operations on the arrays, such as kernel launches and data
+  transfers), and ``sync=False`` should be passed to functions importing arrays
+  through the interface.
+
+
+Lifetime management
+-------------------
+
+Obtaining the value of the ``__cuda_array_interface__`` property of any object
+has no effect on the lifetime of the object from which it was created - in
+particular one should note that the interface does not specify a slot for the
+owner of the data.
+
+As a result, it is imperative for a consumer to retain a reference to the object
+owning the data for as long as the consumer makes use of the data.
+
+
+Lifetime management in Numba
+----------------------------
+
+Numba provides two mechanisms for creating device arrays depending on the
+required effect on the lifetime of the object from which they are created:
+
+- :func:`cuda.as_cuda_array`: This creates a device array that holds a reference
+  to the owning object. As long as a reference to the device array is held, its
+  underlying data will also be kept alive, even if all other references to the
+  original owning object have been dropped.
+- :func:`from_cuda_array_interface`: This creates a device array with no
+  reference to the owning object by default. The owning object, or some other
+  object to be considered the owner can be passed in the ``owner`` parameter.
```

Comment:
`from_cuda_array_interface` looks like a useful function. Didn't realize it was there. Also wasn't able to find it in the docs. Have added issue ( https://github.com/numba/numba/issues/5164 ) to document this.

Reply:
Thanks! I had noticed the links to them in the docs weren't resolving but hadn't looked into it yet...